tags/serializers.py

```diff
@@ -0,0 +1,70 @@
+from rest_framework import serializers
+from rest_framework.relations import PrimaryKeyRelatedField
+
+from tags.models import Tag, Namespace
+
+
+class NamespaceSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Namespace
+        fields = ("id", "name", "scope", "scoped_to")
+
+    def validate_scope(self, value):
+        if value not in Namespace.SCOPES.keys():
+            raise serializers.ValidationError("The scope isn't valid")
+        return value
+
+    def validate(self, data):
+        if len(data) == 0:
+            return data
+
+        if bool(Namespace.SCOPES.get(data.get("scope", None), False)) != bool(
+            data.get("scoped_to", False)
+        ):  # logical XOR
+            raise serializers.ValidationError(
+                "Must define scope and scoped_to together"
+            )
+
+        if Namespace.SCOPES.get(data.get("scope")) and not Namespace.SCOPES.get(
+            data.get("scope")
+        ).objects.get(pk=data.get("scoped_to")):
```

Comment:
IMHO, one thing which could help to read this whole function would be to define the local variables `scope = data.get("scope", None)` and `scoped_to = data.get(…)`